### ISSUE TITLE
Rebalance interrupts after hot-plug (irqbalance)

### ISSUE THREAD

**Issue by c3d** [kata-containers/kata-containers#1195]
**Which feature do you think can be improved?**

The fact that we start with a single CPU and then hot-plug more as we need them may have an adverse side effect with interrupt rebalancing. For example, in a system that started with a single CPU and then had 8 more hot-plugged, we observe that at least for some drivers, multiqueue is disabled during boot (because of the single CPU) and that all interrupts pile on the initial CPU. If we add more initial CPUs, multiqueue is enabled, but interrupts seem to only be dispatched on the original boot CPUs.

**How can it be improved?**

It may be necessary either to run something like `irqbalance` or to perform a similar action in the agent.

**Additional Information**

Here is an example when running with a single CPU at boot:

`dmesg` indicates multi-queue is disabled:
```
[    9.776771] iavf 0000:01:01.0: Multiqueue Disabled: Queue pair count = 1
```

Interrupt dispatch shows a single CPU even after on-lining more:

```
          CPU0       CPU1       CPU2       CPU3       CPU4       CPU5       CPU6       CPU7       CPU8       
 0:         13          0          0          0          0          0          0          0          0  IR-IO-APIC   2-edge      timer
 1:          0          0          0          0          0          0          0          0          0  IR-IO-APIC   1-edge      i8042
 8:          0          0          0          0          0          0          0          0          0  IR-IO-APIC   8-edge      rtc0
 9:          9          0          0          0          0          0          0          0          0  IR-IO-APIC   9-fasteoi   acpi
20:          7          0          0          0          0          0          0          0          0  IR-IO-APIC  20-fasteoi   virtio2
24:          0          0          0          0          0          0          0          0          0  DMAR-MSI   0-edge      dmar0
25:          7          0          0          0          0          0          0          0          0  IR-PCI-MSI 81920-edge      PCIe PME, aerdrv, pciehp
26:          7          0          0          0          0          0          0          0          0  IR-PCI-MSI 98304-edge      PCIe PME, aerdrv, pciehp
27:          7          0          0          0          0          0          0          0          0  IR-PCI-MSI 114688-edge      PCIe PME, aerdrv, pciehp
28:         12          0          0          0          0          0          0          0          0  IR-PCI-MSI 32768-edge      shpchp
29:          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 131072-edge      virtio3-config
30:      10059          0          0          0          0          0          0          0          0  IR-PCI-MSI 131073-edge      virtio3-virtqueues
31:          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 147456-edge      virtio4-config
32:       8544          0          0          0          0          0          0          0          0  IR-PCI-MSI 147457-edge      virtio4-hiprio
33:    4172499          0          0          0          0          0          0          0          0  IR-PCI-MSI 147458-edge      virtio4-requests.0
34:          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 49152-edge      virtio1-config
35:          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 49153-edge      virtio1-control
36:          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 49154-edge      virtio1-event
37:          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 49155-edge      virtio1-request
38:          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 163840-edge      virtio5-config
39:      16511          0          0          0          0          0          0          0          0  IR-PCI-MSI 163841-edge      virtio5-input.0
40:          1          0          0          0          0          0          0          0          0  IR-PCI-MSI 163842-edge      virtio5-output.0
41:          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 16384-edge      virtio0-config
42:        348          0          0          0          0          0          0          0          0  IR-PCI-MSI 16385-edge      virtio0-virtqueues
43:         10          0          0          0          0          0          0          0          0  IR-PCI-MSI 540672-edge      iavf-0000:01:01.0:mbx
44:   17166067          0          0          0          0          0          0          0          0  IR-PCI-MSI 540673-edge      iavf-eth1-TxRx-0

NMI:          0          0          0          0          0          0          0          0          0   Non-maskable interrupts
LOC:     578264     625442     657520     679519     551991     542349     560732     701035     453238   Local timer interrupts
SPU:          0          0          0          0          0          0          0          0          0   Spurious interrupts
PMI:          0          0          0          0          0          0          0          0          0   Performance monitoring interrupts
IWI:         29         24         26         26         23         25         24         23         19   IRQ work interrupts
RTR:          0          0          0          0          0          0          0          0          0   APIC ICR read retries

RES:      58217    1237667     827666     725573     883093     841130    1212502     742799    1002374   Rescheduling interrupts

CAL:       1616       2434       2101       2089       1639       1920       1462       1980       1542   Function call interrupts
TLB:         43       1085        655        651        148        436        243       1160        303   TLB shootdowns
TRM:          0          0          0          0          0          0          0          0          0   Thermal event interrupts
THR:          0          0          0          0          0          0          0          0          0   Threshold APIC interrupts
DFR:          0          0          0          0          0          0          0          0          0   Deferred Error APIC interrupts
MCE:          0          0          0          0          0          0          0          0          0   Machine check exceptions
MCP:          8          8          8          8          8          8          8          8          8   Machine check polls
HYP:          0          0          0          0          0          0          0          0          0   Hypervisor callback interrupts
HRE:          0          0          0          0          0          0          0          0          0   Hyper-V reenlightenment interrupts
HVS:          0          0          0          0          0          0          0          0          0   Hyper-V stimer0 interrupts
ERR:          0
MIS:          0
PIN:          0          0          0          0          0          0          0          0          0   Posted-interrupt notification event
NPI:          0          0          0          0          0          0          0          0          0   Nested posted-interrupt event
PIW:          0          0          0          0          0          0          0          0          0   Posted-interrupt wakeup event
```

Two lines are of particular interest here:

* Line 44 indicates that there is only once CPU that receives all interrupts for the card
* Line `RES` (reschedule interrupts) indicates that this puts some pressure on the rest of the system, in the sense that reschedule interrupts are sent in higher number to all other CPUs, presumably to pull some processes there.

In a case where we start with 4 CPUs, then multiqueue is enabled in the `iavf` driver:

```
[    9.262924] iavf 0000:01:01.0: Multiqueue Enabled: Queue pair count = 4
```

We the observe that interrupts are dispatched only on 4 CPUs:

```
timestamp: 1607109124.861704155
           CPU0       CPU1       CPU2       CPU3       CPU4       CPU5       CPU6       CPU7       CPU8       CPU9       CPU10      CPU11      CPU12      CPU13      CPU14      CPU15      
  0:         11          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0  IR-IO-APIC   2-edge      timer
  1:          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0  IR-IO-APIC   1-edge      i8042
  8:          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0  IR-IO-APIC   8-edge      rtc0
  9:          0          9          0          0          0          0          0          0          0          0          0          0          0          0          0          0  IR-IO-APIC   9-fasteoi   acpi
 20:          0          0          0          0          0          0          7          0          0          0          0          0          0          0          0          0  IR-IO-APIC  20-fasteoi   virtio2
 24:          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0  DMAR-MSI   0-edge      dmar0
 25:          0          0          7          0          0          0          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 81920-edge      PCIe PME, aerdrv, pciehp
 26:          0          0          0          7          0          0          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 98304-edge      PCIe PME, aerdrv, pciehp
 27:          0          0          0          0          7          0          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 114688-edge      PCIe PME, aerdrv, pciehp
 28:          0          0          0          0          0         12          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 32768-edge      shpchp
 29:          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 131072-edge      virtio3-config
 30:          0          0          0      10653          0          0          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 131073-edge      virtio3-virtqueues
 31:          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 147456-edge      virtio4-config
 32:          0          0          0          0          0      10091          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 147457-edge      virtio4-hiprio
 33:          0          0          0          0          0          0    4451829          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 147458-edge      virtio4-requests.0
 34:          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 49152-edge      virtio1-config
 35:          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 49153-edge      virtio1-control
 36:          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 49154-edge      virtio1-event
 37:          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 49155-edge      virtio1-request
 38:          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 163840-edge      virtio5-config
 39:        292          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 163841-edge      virtio5-input.0
 40:          1          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 163842-edge      virtio5-output.0
 41:          0        328          0          0          0          0          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 163843-edge      virtio5-input.1
 42:          0          1          0          0          0          0          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 163844-edge      virtio5-output.1
 43:          0          0        200          0          0          0          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 163845-edge      virtio5-input.2
 44:          0          0          1          0          0          0          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 163846-edge      virtio5-output.2
 45:          0          0          0       4432          0          0          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 163847-edge      virtio5-input.3
 46:          0          0          0          1          0          0          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 163848-edge      virtio5-output.3
 47:          0          0          0          0       2570          0          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 163849-edge      virtio5-input.4
 48:          0          0          0          0          1          0          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 163850-edge      virtio5-output.4
 49:          0          0          0          0          0       4730          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 163851-edge      virtio5-input.5
 50:          0          0          0          0          0          1          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 163852-edge      virtio5-output.5
 51:          0          0          0          0          0          0        946          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 163853-edge      virtio5-input.6
 52:          0          0          0          0          0          0          1          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 163854-edge      virtio5-output.6
 53:          0          0          0          0          0          0          0        991          0          0          0          0          0          0          0          0  IR-PCI-MSI 163855-edge      virtio5-input.7
 54:          0          0          0          0          0          0          0          1          0          0          0          0          0          0          0          0  IR-PCI-MSI 163856-edge      virtio5-output.7
 55:          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 16384-edge      virtio0-config
 56:          0          0          0          0        351          0          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 16385-edge      virtio0-virtqueues
 57:          0          0          0          0          0          0          0          0          9          0          0          0          0          0          0          0  IR-PCI-MSI 540672-edge      iavf-0000:01:01.0:mbx
 58:    4306067          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 540673-edge      iavf-eth1-TxRx-0
 59:          0    5264819          0          0          0          0          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 540674-edge      iavf-eth1-TxRx-1
 60:          0          0    7981262          0          0          0          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 540675-edge      iavf-eth1-TxRx-2
 61:          0          0          0    5418686          0          0          0          0          0          0          0          0          0          0          0          0  IR-PCI-MSI 540676-edge      iavf-eth1-TxRx-3
NMI:          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0   Non-maskable interrupts
LOC:     405353     550805     464993     518880     483788     453491     566400     515918     476062     424938     380603     411671     626643     515951     482642     457176   Local timer interrupts
SPU:          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0   Spurious interrupts
PMI:          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0   Performance monitoring interrupts
IWI:         21         23         21         23         17         19         27         18         20         14         13         11         14         15         17         13   IRQ work interrupts
RTR:          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0   APIC ICR read retries
RES:     430970     730258     166004     458563     357171     452538     499969     208057     469325     293243     287952     449077     335840     542515     447233     400296   Rescheduling interrupts
CAL:       2551       2426       2393       1867       2383       2103       1967       2122       1432       1616       1799       1654       2066       1623       1637       2275   Function call interrupts
TLB:        627        760        825         70        542        241        145        447        132        317        378        187        531        125        129        790   TLB shootdowns
TRM:          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0   Thermal event interrupts
THR:          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0   Threshold APIC interrupts
DFR:          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0   Deferred Error APIC interrupts
MCE:          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0   Machine check exceptions
MCP:          8          8          8          8          8          8          8          8          8          8          8          8          8          8          8          8   Machine check polls
HYP:          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0   Hypervisor callback interrupts
HRE:          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0   Hyper-V reenlightenment interrupts
HVS:          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0   Hyper-V stimer0 interrupts
ERR:          0
MIS:          0
PIN:          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0   Posted-interrupt notification event
NPI:          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0   Nested posted-interrupt event
PIW:          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0          0   Posted-interrupt wakeup event
```

**Comment by fidencio** [comment]
It's a valid issue and as mentioned in the #1365, it does look related. @c3d, mind to take a look at that, and mainly at the comments made by @liubin and see how it affects this one?

**Comment by fidencio** [comment]
I've added the "area/QEMU" and "area/runtime" labels, if you feel those are not the correct ones, please, update them accordingly.
